diff --git a/src/ENCRYPTO_utils/crypto/crypto.cpp b/src/ENCRYPTO_utils/crypto/crypto.cpp
index 4b85bc8..4687d75 100644
--- a/src/ENCRYPTO_utils/crypto/crypto.cpp
+++ b/src/ENCRYPTO_utils/crypto/crypto.cpp
@@ -265,7 +265,9 @@ void crypto::hash_ctr(uint8_t* resbuf, uint32_t noutbytes, uint8_t* inbuf, uint3
 	uint8_t* tmpbuf = (uint8_t*) malloc(ninbytes + sizeof(uint64_t));
 	memcpy(tmpbuf, &ctr, sizeof(uint64_t));
 	memcpy(tmpbuf + sizeof(uint64_t), inbuf, ninbytes);
-	hash_routine(resbuf, noutbytes, tmpbuf, ninbytes+sizeof(uint64_t), sha_hash_buf);
+	uint8_t* hash_buf = (uint8_t*) malloc(get_hash_bytes());
+	hash_routine(resbuf, noutbytes, tmpbuf, ninbytes+sizeof(uint64_t), hash_buf);
+	free(hash_buf);
 	free(tmpbuf);
 }
 
